### PR TITLE
NextPVR UDP Discovery

### PR DESCRIPTION
**Pull request by emveepee**
Added generic UDP discovery to Socket library.  Broadcast to NextPVR on port 16891 to test connected devices.

### CODE REVIEW

**Comment by emveepee** [comment]
NextPVR offers a UDP Discovery model on port 16891,  I  have tested it with Linux and Windows but need to see compiles on other platforms which I cannot build.

@ksooo @phunkyfish Do you know who can review this broadcast approach?

The generic Socket.h/cpp functions came from Mediaportal and have been in NextPVR for years but did not include broadcast logic.  I added these so others may use them.

@sub3 do not merge this there will be much more integration in another update if this PR approach is accepted

**Comment by emveepee** [comment]
As an update confirmed that the UDP discovery works on macos and Android (non root)

I concluded my testing so I have limited this PR to just the changed files for discovery.

**Comment by emveepee** [comment]
I concluded my testing so I have limited this PR to just the changed files for discovery.

File: src/BackendRequest.cpp [hunk 1]
```diff
@@ -97,4 +99,53 @@ namespace NextPVR
...
   }
+  std::vector<std::string> Request::Discovery()
+  {
+    std::vector<std::string> foundAddress;
+    Socket* socket = new Socket(af_inet, pf_inet, sock_dgram, udp);
+    if (socket->create())
+    {
+      bool optResult;
+      int broadcast = 1;
+      if (optResult = socket->SetSocketOption(SOL_SOCKET, SO_REUSEADDR, (char*)&broadcast, sizeof(broadcast)))
```
Comment (phunkyfish) [inline src/BackendRequest.cpp]: C++ cast?

File: src/BackendRequest.cpp [hunk 2]
```diff
@@ -97,4 +99,53 @@ namespace NextPVR
...
+        XBMC->Log(LOG_ERROR, "SO_REUSEADDR %d", optResult);
+      broadcast = 1;
+      if (optResult = socket->SetSocketOption(SOL_SOCKET, SO_BROADCAST, (char*)&broadcast, sizeof(broadcast)))
```
Comment (phunkyfish) [inline src/BackendRequest.cpp]: C++ cast?

File: src/BackendRequest.cpp [hunk 3]
```diff
@@ -97,4 +99,53 @@ namespace NextPVR
...
+        XBMC->Log(LOG_ERROR, "SO_BROADCAST %d", optResult);
+#if defined(TARGET_WINDOWS)
+      DWORD timeout = 5 * 1000;
+      if (optResult = socket->SetSocketOption(SOL_SOCKET, SO_RCVTIMEO, (char*)&timeout, sizeof(timeout)))
```
Comment (phunkyfish) [inline src/BackendRequest.cpp]: C++ cast?

File: src/BackendRequest.cpp [hunk 4]
```diff
@@ -97,4 +99,53 @@ namespace NextPVR
...
+        XBMC->Log(LOG_ERROR, "WINDOWS SO_RCVTIMEO %d", optResult);
+#else
+      struct timeval tv;
+      tv.tv_sec = 5;
+      tv.tv_usec = 0;
+      if (optResult = socket->SetSocketOption( SOL_SOCKET, SO_RCVTIMEO, (char*)&tv, sizeof(tv)))
```
Comment (phunkyfish) [inline src/BackendRequest.cpp]: C++ cast?

**Comment by emveepee** [comment]
@sub3  there have been no comments this week I'd like to go ahead and merge this so I can start using the function in my next commit.

**Comment by phunkyfish** [comment]
I posted a review. You have not commented back yet.

**Comment by emveepee** [comment]
> I posted a review. You have not commented back yet.

Where is that?  I see no reviews showing on the summary

**Comment by phunkyfish** [comment]
Did I forgot to submit it. 1 sec

Sorry another that.

**Review by phunkyfish — COMMENTED** [review]
Just casts which should be C++ casts

File: src/BackendRequest.cpp (hunk 1 above)
Comment (emveepee) [reply src/BackendRequest.cpp]: OK.  Changed and tested in Ubuntu and Windows because of the defines.

File: src/BackendRequest.cpp (hunk 2 above)
Comment (emveepee) [reply src/BackendRequest.cpp]: Changed

File: src/BackendRequest.cpp (hunk 3 above)
Comment (emveepee) [reply src/BackendRequest.cpp]: Changed

File: src/BackendRequest.cpp (hunk 4 above)
Comment (emveepee) [reply src/BackendRequest.cpp]: Changed

**Review by phunkyfish — APPROVED** [review]
👍

**Pull request merged**